Mathlib/Topology/UniformSpace/Ultra/Pseudometrizable.lean

```diff
@@ -0,0 +1,473 @@
+/-
+Copyright (c) 2025 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Data.Real.Archimedean
+import Mathlib.Topology.MetricSpace.BundledFun
+import Mathlib.Topology.UniformSpace.Ultra.Basic
+
+/-!
+# Ultrametric (nonarchimedean) uniform spaces induced by pseudometrics
+
+Ultrametric (nonarchimedean) uniform spaces are such that they are induced by systems
+of pseudometrics having a uniformity based on equivalence relations.
+
+## Main results
+
+* `UniformSpace.pseudoMetrizable.IsUltraUniformity.ofPseudoMetricSystem_pseudoMetricSystem_eq`:
+  Any uniform space has a natural system of pseudometrics definable on it,
+  comprised of those pseudometrics constructed from a descending chain of
+  equivalence relation entourages. In a nonarchimedean uniformity, this pseudometric system
+  induces the uniformity.
+
+## Implementation notes
+
+The proof follows the construction in
+[D. Windisch, *Equivalent characterizations of non-Archimedean uniform spaces*][windisch2021],
+with conditions on the descending chain of equivalence relation entourages simplified
+to easily expose API of the entourages.
+
+-/
+
+open scoped Uniformity
+
+variable {X : Type*}
+
+/-- Any set of pseudometrics can induce a uniform space, where the entourages are
+any open ball of positive radius for any of the pseudometrics. -/
+def UniformSpace.Core.ofPseudoMetricSystem (M : Set (PseudoMetric X ℝ)) :
+    UniformSpace.Core X where
+  uniformity := .generate <| (fun εd ↦ {xy | εd.2 xy.1 xy.2 < εd.1}) '' ((Set.Ioi 0 : Set ℝ) ×ˢ M)
+  refl := by
+    simp only [Filter.principal, idRel_subset, Filter.le_generate_iff, Set.image_subset_iff,
+      Set.preimage_setOf_eq, Set.mem_setOf_eq, PseudoMetric.refl]
+    intro
+    aesop
+  symm := by
+    rw [Filter.tendsto_iff_comap]
+    refine (Filter.generate_image_preimage_le_comap _ _).trans' ?_
+    rw [← Set.image_swap_eq_preimage_swap, Set.image_image, Set.image_swap_eq_preimage_swap]
+    simp [PseudoMetric.symm]
+  comp := by
+    rw [Filter.le_generate_iff]
+    intro s
+    simp only [Set.mem_image, Set.mem_prod, Set.mem_Ioi, Prod.exists, Filter.mem_sets,
+      forall_exists_index, and_imp]
+    rintro ε d εpos hd rfl
+    rw [Filter.mem_lift'_sets (Monotone.compRel _ _)]
+    · refine ⟨{xy | d xy.1 xy.2 < ε / 2}, Filter.mem_generate_of_mem ?_, ?_⟩
+      · simp only [Set.mem_image, Set.mem_prod, Set.mem_Ioi, Prod.exists]
+        exact ⟨ε / 2, d, ⟨by simp [εpos], hd⟩, rfl⟩
+      · intro ⟨a, b⟩
+        rw [mem_compRel]
+        simp only [Set.mem_setOf_eq, forall_exists_index, and_imp]
+        intro c hac hcb
+        refine (d.triangle _ _ _).trans_lt ((add_lt_add hac hcb).trans_le ?_)
+        simp
+    · exact monotone_id
+    · exact monotone_id
```

Comment:
There is a simpler way to write this: each pseudo-metric induced a uniformity thanks to [UniformSpace.ofFun](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/OfFun.html#UniformSpace.ofFun), and then this is just the infimum of the uniformities associated to each pseudo-metric. This should significantly simplify the rest as well.

Reply:
I can't find the `Min` instance on `UniformSpace`, do we have it?

Reply:
[instCompleteLatticeUniformSpace](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/Basic.html#instCompleteLatticeUniformSpace)